### ISSUE TITLE
[XEB] random_quantum_circuit_generation assumes neighboring GridQubits are connected

### ISSUE THREAD

**Issue by mpharrigan** [quantumlib/Cirq#3342]
The random circuits generated take a list of qubit to implicitly define the device. In particular: it assumes that adjacent GridQubits should be interacted. I don't think this has ever happened, but you can imagine having two neighboring qubits that you *don't* want to interact; either because you're trying to model a certain circuit topology or if the coupling is physically non-functional between the qubits.

In the first case, you can argue that that is out-of-scope for a GridInteractionLayer and the second case you could argue it's not a big deal since we haven't run into it yet. Wanted to see whether others would see utility in (optionally) taking a graph (list of edges) during circuit generation

**Comment by balopat** [comment]
Discussed on Cirq Cynque: 
- it sounds like the best place for the responsibility to store the adjacency information would be in `Device` (maybe within a `DeviceTopology` class)

**Comment by mpharrigan** [comment]
Yes, we need to store the adjacency information *and* update the API of the random_quantum_circuit_generation function to take the adjacency information

**Comment by verult** [comment]
@MichaelBroughton Marking this one before-1.0 since it potentially requires Devices changes. Please feel free to change it though if you think it's purely a feature addition.

**Comment by mpharrigan** [comment]
I think devices now support adjacency information, so that part's fine.

The function in question still needs to learn to consume adjacency information, but that can be a post-1.0 feature addition

**Comment by sankalpsthakur** [comment]
I tested the current `main` implementation path at `d7e0f27` and confirmed a small backwards-compatible API direction for this issue:

- add an optional keyword-only `device_graph: nx.Graph | None = None` parameter to `random_rotations_between_grid_interaction_layers_circuit`;
- when provided, restrict its edges to qubits in the requested qubit set and sort the resulting pairs for deterministic seeded output;
- otherwise retain the existing `_coupled_qubit_pairs(qubits)` behavior.

A focused regression using a 2x2 `GridQubit` rectangle with one physically adjacent edge deliberately omitted can assert that the set of emitted two-qubit-operation pairs exactly equals the supplied graph edges. The complete `random_quantum_circuit_generation_test.py` module passes with that behavior (`21 passed`).

This also preserves existing callers because the new argument is keyword-only and defaults to the current implicit grid-adjacency behavior.
